File: Editorial/110_Balanced_Binary_Tree/Go/2_Bottom_up_recursion.go
type TreeInfo struct {
    height   int
    balanced bool
}

func abs(a int) int {
    if a < 0 {
        return -1 * a
    }
    return a
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

// Returns whether the tree at root is balanced, along with the tree's height.
func isBalancedTreeHelper(root *TreeNode) TreeInfo {
    if root == nil {
        return TreeInfo{-1, true}
    }
    left := isBalancedTreeHelper(root.Left)
    if !left.balanced {
        return TreeInfo{-1, false}
    }
    right := isBalancedTreeHelper(root.Right)
    if !right.balanced {
        return TreeInfo{-1, false}
    }
    if abs(left.height-right.height) < 2 {
        return TreeInfo{max(left.height, right.height) + 1, true}
    }
    return TreeInfo{-1, false}
}

func isBalanced(root *TreeNode) bool {
    return isBalancedTreeHelper(root).balanced
}

File: Editorial/0101-0200/0110_Balanced_Binary_Tree/TypeScript/2_Bottom_up_recursion.ts
// Utility class to store information from recursive calls
class TreeInfo {
    constructor(
        public height: number,
        public balanced: boolean,
    ) {}
}
function isBalanced(root: TreeNode | null): boolean {
    // Returns whether the tree at root is balanced, along with the tree's height.
    function isBalancedTreeHelper(root: TreeNode | null): TreeInfo {
        // An empty tree is both balanced and has a height -1.
        if (root == null) {
            return new TreeInfo(-1, true);
        }
        // Checks whether the subtrees are balanced or not.
        const left = isBalancedTreeHelper(root.left);
        if (!left.balanced) {
            return new TreeInfo(-1, false);
        }
        const right = isBalancedTreeHelper(root.right);
        if (!right.balanced) {
            return new TreeInfo(-1, false);
        }
        // The obtained height from recursive calls can also determine
        // that the current node is balanced.
        if (Math.abs(left.height - right.height) < 2) {
            return new TreeInfo(Math.max(left.height, right.height) + 1, true);
        }
        return new TreeInfo(-1, false);
    }
    return isBalancedTreeHelper(root).balanced;
}
